### PR TITLE
[CORE] Unpublish artefacts

### PR DESCRIPTION
**Pull request by cheezus1**
### What does this PR do?
Adds PublisherHandler abstraction, which is to be implemented for every artefact synchronizer.

Adds OData PublisherHandler implementation.

### CODE REVIEW

File: modules/core/core-publisher/src/main/java/org/eclipse/dirigible/core/publisher/api/PublisherHandler.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import org.eclipse.dirigible.core.scheduler.api.SchedulerException;
+import org.eclipse.dirigible.database.persistence.PersistenceManager;
+import org.eclipse.dirigible.database.sql.SqlFactory;
+import org.eclipse.dirigible.repository.api.IRepositoryStructure;
+
+import javax.sql.DataSource;
+import java.sql.Connection;
+import java.sql.SQLException;
+
+public abstract class PublisherHandler {
```
Comment (ThuF) [inline modules/core/core-publisher/src/main/java/org/eclipse/dirigible/core/publisher/api/PublisherHandler.java:24]: Create `IPublishHandler` interface to extract the abstract methods there:

```java
public abstract class PublisherHandler {
```

->

```java
public abstract class AbstractPublishHandler implements IPublishHandler {
```

File: modules/core/core-publisher/src/main/java/org/eclipse/dirigible/core/publisher/api/PublisherHandler.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    private DataSource dataSource = (DataSource) StaticObjects.get(StaticObjects.SYSTEM_DATASOURCE);
+
+    public abstract void beforePublish(String location) throws SchedulerException;
+
+    public abstract void afterPublish(String location) throws SchedulerException;
+
+    public abstract void beforeUnpublish(String location) throws SchedulerException;
+
+    public abstract void afterUnpublish(String location) throws SchedulerException;
```
Comment (ThuF) [inline modules/core/core-publisher/src/main/java/org/eclipse/dirigible/core/publisher/api/PublisherHandler.java:38]: Extract into `IPublishHandler` interface

File: modules/engines/engine-odata/src/main/resources/META-INF/services/org.eclipse.dirigible.core.publisher.api.PublisherHandler [hunk 3]
```diff
@@ -0,0 +1 @@
+org.eclipse.dirigible.engine.odata2.synchronizer.ODataSynchronizerPublisherHandler	# OData Synchronizer Publisher Handler
```
Comment (ThuF) [inline modules/engines/engine-odata/src/main/resources/META-INF/services/org.eclipse.dirigible.core.publisher.api.PublisherHandler:1]: Change the file name to point to the`IPublishHandler` interface:


`org.eclipse.dirigible.core.publisher.api.PublisherHandler `

->

`org.eclipse.dirigible.core.publisher.api.IPublishHandler`

File: modules/core/core-publisher/src/main/java/org/eclipse/dirigible/core/publisher/api/PublisherHandler.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    protected String getLocationQueryParam(String location, boolean includeLeadingSeparator) {
```
Comment (ThuF) [inline modules/core/core-publisher/src/main/java/org/eclipse/dirigible/core/publisher/api/PublisherHandler.java:40]: Change the method visibility from `protected` to `private`

File: modules/core/core-publisher/src/main/java/org/eclipse/dirigible/core/publisher/api/PublisherHandler.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            SqlFactory sqlFactory = SqlFactory.getNative(connection);
+
+            if (sqlFactory.exists(connection, table)) {
+                String sql = sqlFactory.delete().from(table).where(new StringBuilder().append(column).append(" LIKE ?").toString()).build();
+                persistenceManager.execute(connection, sql, locationQueryParam);
+            }
+        } catch (SQLException e) {
+            throw new SchedulerException(e);
+        }
+    }
```
Comment (ThuF) [inline modules/core/core-publisher/src/main/java/org/eclipse/dirigible/core/publisher/api/PublisherHandler.java:65]: Change the method in the following way:
```java
    protected void removeMetadata(PersistenceManager persistenceManager, String table, String column, String location, boolean includeLeadingSeparator) throws SchedulerException {
        try (Connection connection = dataSource.getConnection()) {
            SqlFactory sqlFactory = SqlFactory.getNative(connection);

            if (sqlFactory.exists(connection, table)) {
                String sql = sqlFactory.delete().from(table).where(new StringBuilder().append(column).append(" LIKE ?").toString()).build();
                String locationQueryParam = getLocationQueryParam(location, includeLeadingSeparator);
                persistenceManager.execute(connection, sql, locationQueryParam);
            }
        } catch (SQLException e) {
            throw new SchedulerException(e);
        }
    }
```

File: modules/engines/engine-odata/src/main/java/org/eclipse/dirigible/engine/odata2/synchronizer/ODataSynchronizerPublisherHandler.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+    @Override
+    public void afterUnpublish(String location) throws SchedulerException {
+        String locationQueryParam = getLocationQueryParam(location, true);
+
+        removeMetadata(new PersistenceManager<ODataDefinition>(), "DIRIGIBLE_ODATA", "ODATA_LOCATION", locationQueryParam);
+        removeMetadata(new PersistenceManager<ODataContainerDefinition>(), "DIRIGIBLE_ODATA_CONTAINER", "ODATAC_LOCATION", locationQueryParam);
+        removeMetadata(new PersistenceManager<ODataHandlerDefinition>(), "DIRIGIBLE_ODATA_HANDLER", "ODATAH_LOCATION", locationQueryParam);
+        removeMetadata(new PersistenceManager<ODataMappingDefinition>(), "DIRIGIBLE_ODATA_MAPPING", "ODATAM_LOCATION", locationQueryParam);
+        removeMetadata(new PersistenceManager<ODataSchemaDefinition>(), "DIRIGIBLE_ODATA_SCHEMA", "ODATAX_LOCATION", locationQueryParam);
```
Comment (ThuF) [inline modules/engines/engine-odata/src/main/java/org/eclipse/dirigible/engine/odata2/synchronizer/ODataSynchronizerPublisherHandler.java:44]: Update to:
```java
removeMetadata(new PersistenceManager<ODataDefinition>(), "DIRIGIBLE_ODATA", "ODATA_LOCATION", location, true);
removeMetadata(new PersistenceManager<ODataContainerDefinition>(), "DIRIGIBLE_ODATA_CONTAINER", "ODATAC_LOCATION", location, true);
removeMetadata(new PersistenceManager<ODataHandlerDefinition>(), "DIRIGIBLE_ODATA_HANDLER", "ODATAH_LOCATION", location, true);
removeMetadata(new PersistenceManager<ODataMappingDefinition>(), "DIRIGIBLE_ODATA_MAPPING", "ODATAM_LOCATION", location, true);
removeMetadata(new PersistenceManager<ODataSchemaDefinition>(), "DIRIGIBLE_ODATA_SCHEMA", "ODATAX_LOCATION", location, true);
```

**Review by ThuF — CHANGES_REQUESTED** [review]
Few minor comments to be addressed, before merging the PR

**Review by delchev — APPROVED** [review]

**Pull request merged**